databroker/src/broker.rs

```diff
@@ -1530,6 +1592,278 @@ impl<'a, 'b> AuthorizedAccess<'a, 'b> {
+            let actuation_changes = actuation_change_per_vss_id.1.clone();
+
+            let opt_actuation_subscription = actuation_subscriptions
+                .iter()
+                .find(|subscription| subscription.vss_ids.contains(&vss_id));
+            match opt_actuation_subscription {
+                Some(actuation_subscription) => {
+                    let is_expired = actuation_subscription.permissions.expired();
+                    match is_expired {
+                        Err(_) => {
+                            let message = format!(
+                                "Permission for vss_ids {:?} expired",
+                                actuation_subscription.vss_ids
+                            );
+                            return Err((ActuationError::PermissionExpired, message));
+                        }
+                        Ok(_) => {
+                            if !actuation_subscription.actuation_provider.is_available() {
+                                let message =
+                                    format!("Provider for vss_id {} does not exist", vss_id);
+                                return Err((ActuationError::ProviderNotAvailable, message));
+                            }
+
+                            actuation_subscription
+                                .actuation_provider
+                                .actuate(actuation_changes)
+                                .await?
+                        }
+                    }
+                }
+                None => {
+                    let message = format!("actuation provider for vss_id {} not available", vss_id);
+                    return Err((ActuationError::ProviderNotAvailable, message));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn actuate(
+        &self,
+        vss_id: &i32,
+        data_value: &DataValue,
+    ) -> Result<(), (ActuationError, String)> {
+        let vss_id = *vss_id;
+        let result_entry = self.get_entry_by_id(vss_id).await;
+        match result_entry {
+            Ok(entry) => {
+                let vss_path = entry.metadata.path;
+                let result_can_write_actuator =
+                    self.permissions.can_write_actuator_target(&vss_path);
+                match result_can_write_actuator {
+                    Ok(_) => {}
+                    Err(PermissionError::Denied) => {
+                        let message = format!("Permission denied for vss_path {}", vss_path);
+                        return Err((ActuationError::PermissionDenied, message));
+                    }
+                    Err(PermissionError::Expired) => {
+                        return Err((
+                            ActuationError::PermissionExpired,
+                            "Permission expired".to_string(),
+                        ))
+                    }
+                }
+            }
+            Err(ReadError::NotFound) => {
```

Comment:
This is the third time we have the code snippet below. I have not checked the code structure that thoroughly, but it would be possible and feasible to avoid code duplication by using a helper method?

```
Err(ReadError::NotFound) => {
                    let message = format!("Could not resolve vss_path of vss_id {}", vss_id);
                    return Err((ActuationError::NotFound, message));
                }
```

Reply:
Could you elaborate here? 
The ActuationError-specific error messages make it kinda hard to separate this into a separate method. I could add a ReadError -> ActuationError conversion (same obviously for a PermissionError -> ActuationError), but I would still need to provide the error message as a parameter so it contains useful data like vss_id or vss_path. Therefore I don't see a lot of gain.

Reply:
We have calls to             `self.get_entry_by_id(vss_id).await;` on line 1585, 1689 and 1772, i.e. in all 3 actuate methods, so it was just a thought if we could reuse that part by adding the call and the common error handling in a helper method. Something like

```
result = self.MyActuateHelperMethod
match result {
  Ok(res) => ... // Do something
  Err(err)  => return Err(err), // Reuse the error created by helper method
}

```

Reply:
Good point, done! 
Encapsulated the logic into a separate "can_write_actuator_target" method, which takes a vss_id, looks it up and checks if permissions to write them are available. Really saved a few lines of code.